pkg/solana/logpoller/loader_test.go

```diff
@@ -53,42 +58,204 @@ func TestEncodedLogCollector_ParseSingleEvent(t *testing.T) {
+		RunAndReturn(latestBlockhashReturnFunc(&latest))
 
-	client.EXPECT().GetBlockWithOpts(mock.Anything, slot, mock.Anything).Return(&rpc.GetBlockResult{
-		Transactions: []rpc.TransactionWithMeta{
-			{
-				Meta: &rpc.TransactionMeta{
-					LogMessages: messages,
-				},
-			},
-		},
-		Signatures:  []solana.Signature{sig},
-		BlockHeight: &blockHeight,
-	}, nil).Twice()
+	client.EXPECT().
+		GetBlocks(
+			mock.Anything,
+			mock.MatchedBy(getBlocksStartValMatcher),
+			mock.MatchedBy(getBlocksEndValMatcher(&latest)),
+			rpc.CommitmentFinalized,
+		).
+		RunAndReturn(getBlocksReturnFunc(false))
+
+	client.EXPECT().
+		GetBlockWithOpts(mock.Anything, mock.Anything, mock.Anything).
+		RunAndReturn(func(_ context.Context, slot uint64, _ *rpc.GetBlockOpts) (*rpc.GetBlockResult, error) {
+			height := slot - 1
+
+			result := rpc.GetBlockResult{
+				Transactions: []rpc.TransactionWithMeta{},
+				Signatures:   []solana.Signature{},
+				BlockHeight:  &height,
+			}
+
+			_, _ = rand.Read(result.Blockhash[:])
+
+			if slot == 42 {
+				var sig solana.Signature
+				_, _ = rand.Read(sig[:])
+
+				result.Signatures = []solana.Signature{sig}
+				result.Transactions = []rpc.TransactionWithMeta{
+					{
+						Meta: &rpc.TransactionMeta{
+							LogMessages: messages,
+						},
+					},
+				}
+			}
+
+			return &result, nil
+		})
 
 	tests.AssertEventually(t, func() bool {
 		return parser.Called()
 	})
+}
+
+func TestEncodedLogCollector_MultipleEventOrdered(t *testing.T) {
+	t.Parallel()
+
+	client := new(mocks.RPCClient)
+	parser := new(testParser)
+	ctx := tests.Context(t)
+
+	collector := logpoller.NewEncodedLogCollector(client, parser, logger.Nop())
+
+	require.NoError(t, collector.Start(ctx))
+	t.Cleanup(func() {
+		require.NoError(t, collector.Close())
+	})
+
+	var latest atomic.Uint64
+
+	latest.Store(uint64(40))
+
+	slots := []uint64{44, 43, 42, 41}
+	sigs := make([]solana.Signature, len(slots))
+	hashes := make([]solana.Hash, len(slots))
+
+	for idx := range len(sigs) {
+		_, _ = rand.Read(sigs[idx][:])
+		_, _ = rand.Read(hashes[idx][:])
+	}
+
+	client.EXPECT().
+		GetLatestBlockhash(mock.Anything, rpc.CommitmentFinalized).
+		RunAndReturn(latestBlockhashReturnFunc(&latest))
+
+	client.EXPECT().
+		GetBlocks(
+			mock.Anything,
+			mock.MatchedBy(getBlocksStartValMatcher),
+			mock.MatchedBy(getBlocksEndValMatcher(&latest)),
+			rpc.CommitmentFinalized,
+		).
+		RunAndReturn(getBlocksReturnFunc(false))
+
+	client.EXPECT().
+		GetBlockWithOpts(mock.Anything, mock.Anything, mock.Anything).
+		RunAndReturn(func(_ context.Context, slot uint64, _ *rpc.GetBlockOpts) (*rpc.GetBlockResult, error) {
+			slotIdx := -1
+			for idx, slt := range slots {
+				if slt == slot {
+					slotIdx = idx
+
+					break
+				}
+			}
+
+			// imitate loading block data out of order
+			if slot == 43 || slot == 41 {
+				// force slot 43 to return after 44
+				// same for 41 after 42
+				time.Sleep(1 * time.Second)
```

Comment:
I think technically it can't cause a flake, worst case they would end up in order instead of out of order so it wouldn't be fully testing the ordering on that run. But still, better to avoid relying on a Sleep for testing this